### ISSUE TITLE
wrong result: Query with subquery, aggregate, outer join wrongly reduced to Constant 

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#7047]
### What version of Materialize are you using?

```
v0.8.1-dev (516a132e2)
```

### What was the issue?

This reasonably valid, plausible, realistic, non-impossible, entirely possible, non-self-defeating query:

```

drop table if exists t1 cascade;
drop table if exists t2 cascade;

create table t1 (f1 integer);
insert into t1 values (0);
insert into t1 values (1);

SELECT * FROM (
	SELECT 123, COUNT(right_table.f1) AS aggregate
	FROM t1 AS left_table
	LEFT JOIN t1 AS right_table
	ON FALSE
	GROUP BY 1
) AS subquery,
t1 AS outer_table 
WHERE outer_table.f1 = subquery.aggregate;
```

Is judged to have a Constant plan:

```
 Optimized Plan 
----------------
 %0 =          +
 | Constant    +
```

and returns no rows whereas it should have returned a row as per Postgresql:

```
 ?column? | aggregate | f1 
----------+-----------+----
      123 |         0 |  0
```

I think the ```Constant``` verdict comes from the ```ON FALSE``` clause, however this is incorrect, as an outer join can produce records even in the face of ```ON FALSE```. The subquery taken alone does not have a Constant plan, but it also does not do a Join either:

```
materialize=> explain SELECT 123, COUNT(right_table.f1) AS aggregate
materialize->  FROM t1 AS left_table
materialize->  LEFT JOIN t1 AS right_table
materialize->  ON FALSE
materialize->  GROUP BY 1;
           Optimized Plan            
-------------------------------------
 %0 =                               +
 | Get materialize.public.t1 (u3687)+
 | Reduce group=()                  +
 | | agg count(null)                +
 | Map 123                          +
 | Project (#1, #0)                 +
```

### Is the issue reproducible? If so, please provide reproduction instructions.

```
drop table if exists t1 cascade;

create table t1 (f1 integer);
insert into t1 values (0);

SELECT * FROM (
	SELECT 123, COUNT(right_table.f1) AS aggregate
	FROM t1 AS left_table
	LEFT JOIN t1 AS right_table
	ON FALSE
	GROUP BY 1
) AS subquery,
t1 AS outer_table 
WHERE outer_table.f1 = subquery.aggregate;
```

**Comment by philip-stoev** [comment]
Yes.  but the plan is now inner-join-y whereas the original query has an outer join. Would you mind pushing this query into your .slt test and closing this ticket as a duplicate?

**Comment by asenac** [comment]
Done. The inner join plan is OK, since `count(right_table.f1) on false` is equivalent to `count(null)`.

However, we should detect that `count(null)` is always 0. Filing a ticket for that.

This issue will be automatically closed when merging #7052 .
